curriculum/challenges/english/25-front-end-development/lecture-understanding-form-validation/6733aae9d25004f60d1e86f2.md

```diff
@@ -10,6 +10,140 @@ dashedName: what-are-some-ways-to-validate-forms-using-javascript
+
+But how can you actually use it?
+
+Let's say you wanted employees at a company to send feedback messages through a form like this:
+
+```html
+<form>
+  <label>Enter your email: </label>
+  <input required type="email" />
+
+  <label>Enter your feedback: </label>
+  <textarea required placeholder="Your feedback here..."></textarea>
+
+  <button type="submit">Submit Feedback</button>
+</form>
+```
+
+We are using the email `input` which comes with built in validation to check for basic validation like if the `input` includes the at(`@`) sign.
+
+But what if the user provides an email address like `example@email.com`?
+
+This would pass the basic validation, but we want to be more specific about accepting emails from those with a company email address.
+
+This is where we can use the `pattern` attribute to specify that the email address must end in a company email address.
+
+Here is what the updated example will look like:
+
+```html
+<form>
+  <label>Enter your email: </label>
+  <input required placeholder="username@sampleCompany.com" type="email" pattern=".+@sampleCompany\.com" />
+
+  <label>Enter your feedback: </label>
+  <textarea required placeholder="Your feedback here..."></textarea>
+
+  <button type="submit">Submit Feedback</button>
+</form>
+```
+
+Now, if you try to submit the feedback, you will see a message saying "Please match the requested format."
+
+Even though the `input` does have placeholder text showing them the desired format, it would be better to also include a customized error message using JavaScript.
+
+Let's first take a look at the `checkValidity()` method:
+
+```js
+const input = document.querySelector("input");
+
+input.addEventListener("input", (e) => {
+  console.log(e.target.checkValidity())
+})
+```
+
+In the above example, we've queried our input from the DOM, and added an input event listener.
+
+We know that `e.target` refers to the element that triggered the event. In this case, our `input`. But what is the `checkValidity()` method?
+
+This is part of the Constraint Validation API. The `checkValidity()` method returns `true` if the element matches all HTML validation (based on its attributes), and `false` if it fails.
+
+When we try with an invalid input, we see `false` gets logged in the console. Now that we know the input is invalid, let's report the invalidity:
+
+```js
+const input = document.querySelector("input");
+
+input.addEventListener("input", (e) => {
+  if (!e.target.checkValidity()) {
+    e.target.reportValidity();
+  }
+})
+```
+
+And as a result, you will see the browser's error message "Please match the requested format."
+
+It reports the invalid state immediately, instead of waiting for us to submit the form. But it's still using the default message. This is because the `reportValidity` method only tells the browser that the input is invalid. The browser still chooses how to display *why* it's invalid. That's where the `setCustomValidity` method comes in.
```

Comment:
```suggestion
It reports the invalid state immediately, instead of waiting for us to submit the form. But it's still using the default message. This is because the `reportValidity` method only tells the browser that the input is invalid. The browser still chooses how to display why it's invalid. That's where the `setCustomValidity` method comes in.
```